### PR TITLE
Add recommendation for final await in an async functions

### PR DESCRIPTION
**Pull request by joeytwiddle**
Additional to #669 

I realised just too late that there is another style we can document.

I have seen this pattern in your code and I do think it helps to reduce ambiguity.

(It might have a small performance overhead, consuming a promise only to immediately create another when returning, but I guess that's only a concern for inner loops.)

### CODE REVIEW

**Review by shimks — APPROVED** [review]

**Review by b-admike — APPROVED** [review]

File: pages/en/contrib/style-guide.md [hunk 1]
```diff
@@ -449,6 +449,44 @@ async function getDodo() {
 }
 ```
 
+### Always await the last promise in a function
+
+If the last line of a function handles a promise, always await that promise, regardless whether the result will be returned or not. (This makes it clear whether we did or did not intend to return the value, and dismisses any idea that we are still linking up the promise chain in pre-await style.)
+
```
Comment (raymondfeng) [inline pages/en/contrib/style-guide.md]: Is it better to rephrase it as `This makes it clear whether we intend to return the value or not, and dismisses any idea that we are still linking up the promise chain in pre-await style.`?

File: pages/en/contrib/style-guide.md [hunk 2]
```diff
@@ -449,6 +449,44 @@ async function getDodo() {
...
+  await openCarPark();
+  return openDoors();
+}
+```
+
+**Bad**:
+
+```js
+async function getAnimal() {
+  return getElephant();
```
Comment (raymondfeng) [inline pages/en/contrib/style-guide.md:488]: In this case, `async` keyword is not needed as no `await` is used in the method body.

File: pages/en/contrib/style-guide.md [hunk 3]
```diff
@@ -449,6 +449,44 @@ async function getDodo() {
...
+  return await getElephant();
+}
+```
+
+**Bad**:
+
+```js
+async function openCircus() {
+  await openCarPark();
+  return openDoors();
```
Comment (raymondfeng) [inline pages/en/contrib/style-guide.md:480]: Should we mention this style is not friendly to `try...catch`. For example:
```
async function openCircus() {
  try {
    await openCarPark();
    return await openDoors();
  } catch(err) {
    // rejected promise from openDoors will be trapped here
    throw err;
  }
```

```
async function openCircus() {
  try {
    await openCarPark();
    return openDoors();
  } catch(err) {
    // rejected promise from openDoors will NOT be trapped here
    throw err;
  }
```

File: pages/en/contrib/style-guide.md (hunk 2 above)
Comment (shimks) [reply pages/en/contrib/style-guide.md:488]: I think the point @joeytwiddle was trying to give a bad example of using promises and async/await. Users might be tempted to return a promise for the purpose of promise chaining, so this example is showing them to not do that

Comment (raymondfeng) [reply pages/en/contrib/style-guide.md:488]: I understand. But I would like to help users understand why it's bad.

Comment (joeytwiddle) [reply pages/en/contrib/style-guide.md:488]: In this example I want to emphasize that the problem is the lack of `await`.

If we have an example without `async` I am worried the reader may think that the missing `async` is the problem.

It's worth noting that two sections above we have said "Always use the `async` keyword" so not having `async` is already considered undesirable.

File: pages/en/contrib/style-guide.md (hunk 3 above)
Comment (joeytwiddle) [reply pages/en/contrib/style-guide.md:480]: Thanks, that is another good reason. I have added a note about that.

Although I didn't include specific examples, concerned they would be a lot for the reader to read.

File: pages/en/contrib/style-guide.md (hunk 1 above)
Comment (joeytwiddle) [reply pages/en/contrib/style-guide.md]: Thanks for your clearer rewording. Updated.

File: pages/en/contrib/style-guide.md (hunk 2 above)
Comment (raymondfeng) [reply pages/en/contrib/style-guide.md:488]: That's fine.

**Comment by dhmlau** [comment]
@raymondfeng , are you good with the latest changes in the PR?  
@joeytwiddle, you'd need to rebase the PR.  Thanks.

**Comment by joeytwiddle** [comment]
Rebased it!

**Review by raymondfeng — APPROVED** [review]

**Pull request merged**